lib/ecto/query/planner.ex

```diff
@@ -1042,16 +1042,23 @@ defmodule Ecto.Query.Planner do
 
   defp validate_and_increment(:with_cte, query, with_expr, counter, _operation, adapter) do
     fun = &validate_and_increment(&1, &2, &3, &4, :all, adapter)
-
+    
     {queries, counter} =
       Enum.reduce with_expr.queries, {[], counter}, fn
         {name, opts, %Ecto.Query{} = inner_query}, {queries, counter} ->
+          operation = case opts[:operation] do
+            :update_all -> :update_all
+            :delete_all -> :delete_all
+            :insert_all -> :insert_all
+            _ -> :all
+          end
+
           inner_query = put_in(inner_query.aliases[@parent_as], query)
 
           # We don't want to use normalize_subquery_select because we are
           # going to prepare the whole query ourselves next.
           {_, _, inner_query} = rewrite_subquery_select_expr(inner_query, true)
-          {inner_query, counter} = traverse_exprs(inner_query, :all, counter, fun)
+          {inner_query, counter} = traverse_exprs(inner_query, operation, counter, fun)
```

Comment:
Can we write this as:

```suggestion
          {inner_query, counter} = traverse_exprs(inner_query, opts.operation, counter, fun)
```

And then we make sure it is normalized to `:all` if `nil` is given when created?

Reply:
yes, this makes much more sense. Updated in the latest commit